pkg/templating/manifest_generator.go

```diff
@@ -0,0 +1,306 @@
+		`destination.labels["service-mesh.3scale.net"] == "true"`,
+		fmt.Sprintf(`destination.labels["service-mesh.3scale.net/uid"] == "%s"`, cg.GetUID()),
+	}
+	cg.Rule.Conditions = append(cg.Rule.Conditions, conditions...)
+}
+
+// Returns the hostname for the ConfigGenerator
+func (cg ConfigGenerator) GetHost() string {
+	return cg.host
+}
+
+// Returns the unique UID for the ConfigGenerator
+func (cg ConfigGenerator) GetUID() string {
+	return cg.uid
+}
+
+// Returns the uService ID for the ConfigGenerator
+func (cg ConfigGenerator) GetServiceID() string {
+	return cg.Handler.ServiceID
+}
+
+// GenerateAuthenticationAttributes - templating accessible function used to
+// format the authentication behaviour
+func (i Instance) GenerateAuthenticationAttributes(an AuthenticationMethod) string {
+	var useThis string
+
+	switch an {
+	case ApiKey:
+		useThis = i.generateApiKeyAttributes()
+	case ApplicationID:
+		useThis = i.generateApplicationIdAttributes()
+	case Hybrid:
+		useThis = i.generateApiKeyAttributes() + "\n      " + i.generateApplicationIdAttributes()
+	default:
+		panic("unknown field passed to string generator")
+	}
+
+	return useThis
+}
+
+// yaml generator for api key attribute
+func (i Instance) generateApiKeyAttributes() string {
+	return fmt.Sprintf(`user: %s | ""`, i.formatCredentialsLocation(i.ApiKeyLabel))
+}
+
+// yaml generator for app id/ app key attributes
+func (i Instance) generateApplicationIdAttributes() string {
+	return fmt.Sprintf("properties:\n        %s\n        %s",
+		fmt.Sprintf(`%s: %s | ""`, defaultAppIdLabel, i.formatCredentialsLocation(i.AppIDLabel)),
+		fmt.Sprintf(`%s: %s | ""`, defaultAppKeyLabel, i.formatCredentialsLocation(i.AppKeyLabel)))
+}
+
+// formats an authentication attribute into appropriate string based on the credentials location
+// if no credentials location set, defaults to checking both query params and headers
+func (i Instance) formatCredentialsLocation(key string) string {
+	var formatted string
+
+	switch i.CredentialsLocation {
+	case QueryParams:
+		formatted = fmt.Sprintf(`request.query_params["%s"]`, key)
+	case Headers:
+		formatted = fmt.Sprintf(`request.headers["%s"]`, formatHeaderLabel(key))
+	default:
+		//Unspecified
+		formatted = fmt.Sprintf(`request.query_params["%s"] | request.headers["%s"]`, key, formatHeaderLabel(key))
+	}
+	return formatted
+}
+
+// GenerateListenString - creates a string from the provided Handler replacing unset/invalid values
+// with internal defaults.
+func (h Handler) GenerateListenString() string {
+	if h.ListenAddress == "" {
+		h.ListenAddress = defaultListenAddress
+	}
+
+	if h.Port == 0 {
+		h.Port = defaultListenPort
+	}
+	return fmt.Sprintf("%s:%d", h.ListenAddress, h.Port)
+}
+
+// ConditionsToMatchString returns a valid expression for Istio match condition
+func (r Rule) ConditionsToMatchString() string {
+	var matchOn string
+	for i, condition := range r.Conditions {
+		matchOn += condition
+		if i < len(r.Conditions)-1 {
```

Comment:
you can extract this `len()` call outside the loop - go can't guarantee other threads are not modifying `r.Conditions` while you loop, so it has to keep looking at its length even though it's meant to be constant, so storing it in a variable outside is a good practice (plus you also get to save the subtraction).

Reply:
Yep good spot